File: exam.c
/*
Write a function that displays a user id and a password for a user based on the first
name, last name, and year of birth of the user. The user id is the first letter of the first
name followed by the last name. The password is the first three letters of the first name,
followed by the last two digits of the year of birth, followed by the last three letters of the
last name. The function prototype is void id_password(char first[], char
last[], int year). The first and last names are strings of length at least three. The
year is a four digit integer. The function displays the id and password.
*/
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
void id_password(char first[], char last[], int year){
    char firstthree[4];
    char lastthree[4];
    int i;
    for(i = 0; i < 3; i++){
        firstthree[i] = first[i];
    }
    firstthree[3] = '\0';

    for(i = 0; i < 3; i++){
        lastthree[i] = last[strlen(last) - 3 + i];
    }
    lastthree[3] = '\0';


    printf("%s%02d%s", firstthree, year % 100, lastthree);
}
/*
Write a function that reads an input file and creates a sorted output file. The function
prototype is void file_sort(char *infile, char *outfile). The
function takes the file names as parameters. The input file contains student information
and has the following format. The first line contains the number of students in the file.
Each subsequent line contains information about one student. Each line contains three
fields namely student id, grade, and gpa in that order separated by blanks. The student id
is a positive integer. The grade is a character. The gpa is a double type value. The
function sorts the student information in the ascending order of student id. The ordered
student information is written to the output file. The output file has the same format as
the input file. The function dynamically allocates three arrays and stores the student
information into the arrays. The three arrays are simultaneously sorted. The sorted array
information is written to the output file. The dynamic arrays are freed when the function
completes its work. Structures should not be used
*/
void file_sort(char *infile, char *outfile)
{
    FILE *fp;
    int students;
    fp = fopen(infile, "r");
    fscanf(fp, "%d", &students);
    fgetc(fp);

    int *student_ids;
    student_ids = (int *)malloc(students * sizeof(int));
    char *student_grades;
    student_grades = (char *)malloc(students * sizeof(char));
    double *student_gpas;
    student_gpas = (double *)malloc(students * sizeof(double));


    int i;
    for(i = 0; i < students; i++){
        fscanf(fp, "%d %c %lf", &student_ids[i], &student_grades[i], &student_gpas[i]);
    }
    fclose(fp);

    int j;
    for (i = 0; i < students; i++){
        for (j = i + 1; j < students; j++)
        {
            if (student_ids[j] < student_ids[i])
            {
                int tempid = student_ids[i];
                student_ids[i] = student_ids[j];
                student_ids[j] = tempid;

                char tempgrade = student_grades[i];
                student_grades[i] = student_grades[j];
                student_grades[j] = tempgrade;

                double tempgpa = student_gpas[i];
                student_gpas[i] = student_gpas[j];
                student_gpas[j] = tempgpa;
            }
        }
    }

    FILE *fwrite;
    fwrite = fopen(outfile,"w");
    fprintf(fwrite, "%d\n", students);
    for (i = 0; i < students; i++){
        fprintf(fp, "%d %c %.2f\n", student_ids[i], student_grades[i], student_gpas[i]);
    }
    fclose(fwrite);

    free(student_ids);
    free(student_grades);
    free(student_gpas);


}
int main(){
    // char first[40];
    // char last[40];
    // int year;
    // printf("Enter first name: ");
    // scanf("%s", first);
    // printf("Enter last name: ");
    // scanf("%s", last);
    // printf("enter birth year: ");
    // scanf("%d", &year);
    // id_password(first, last, year);
    file_sort("case9in.txt", "output.txt");
    return 0;
}

File: exam.py
#Write a function sort(list) that sorts a given list of numbers. The function returns
#the sorted list as return value. The given input list remains unchanged. Selection sort must
#be used. Built in sort function cannot be used.

def sort(list):
    for k in range(0, len(list), 1):
        for j in range(k + 1, len(list), 1):
            if list[j] < list[k]:
                temp = list[j]
                list[j] = list[k]
                list[k] = temp
#Write a function that automatically generates the user id and the password for the user.
#The user id and the pass words are based on the user’s last name and first name. The user
#id is the first letter of the first name followed by the last name. The password is the first
#letter of the first name followed the last letter of the first name followed by the first three
#letters of the last name followed by the length of the first name followed by the length of
#the last name. The user may enter the first and last names in lower or upper case letters.
#Both the user id and passwords must be in upper case. For example, if the first name is
#John and the last name is Maxwell then the user id is JMAXWELL and the password is
#JNMAX47. The function id_password(first, last) takes first name and last
#name as inputs and returns user id and password as outputs.


def id_password(first, last):
    userid = (first[0] + last).upper()
    userpassword = (first[0] + first[len(first) - 1] + last[0:3] + str(len(first)) + str(len(last))).upper()
    return userid, userpassword

# Write a function that reads an input file and creates a sorted output file. The function is
# file_sort(infile, outfile). The function takes the file names as parameters.
# The input file contains student information and has the following format. The first line
# contains the number of students in the file. Each subsequent line contains information
# about one student. Each line contains three fields namely student id, name, and gpa in
# that order separated by blank. The student id is a positive integer. The name is a single
# word string. The gpa is a double type value. The function sorts the student information in
# the ascending order of student id. The ordered student information is written to the output
# file. The output file has the same format as the input file. Built in sort function cannot be
# used.

def file_sort(infile, outfile):
    in_file = open(infile, 'r')
    students = int(in_file.readline().strip())

    student_id = []
    student_name = []
    student_gpa = []

    for line in in_file:    
        student_info = line.strip().split()

        student_id = student_id + [student_info[0]]
        student_name = student_name + [student_info[1]]
        student_gpa = student_gpa + [student_info[2]]

    in_file.close()       

    for k in range(0, students, 1):
        for j in range(k + 1, students, 1):
            if int(student_id[j]) < int(student_id[k]):
                temp_id = student_id[j]
                student_id[j] = student_id[k]
                student_id[k] = temp_id

                temp_name = student_name[j]
                student_name[j] = student_name[k]
                student_name[k] = temp_name

                temp_gpa = student_gpa[j]
                student_gpa[j] = student_gpa[k]
                student_gpa[k] = temp_gpa
    
    out_file = open(outfile, 'w')
    out_file.write(str(students) + "\n")

    for k in range (0, students, 1):
        out_file.write(str(student_id[k]) + " " + student_name[k] + " " + str(student_gpa[k]) + "\n")
    
    out_file.close()

class rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width
    
    def getLength(self):
        return self.length
    
    def setLength(self, length):
        self.length = length
    
    def getArea(self):
        return self.length * self.width
    
    def __str__(self):
        return str(self.length) + " " + str(self.width)
    




list = [12,6,3,8,9,1]
list2 = list.copy()
sort(list2)
print(list2)

First = input("Enter your first name: ")
Last = input("Enter your last name: ")
id, password = id_password(First, Last)
print(id, password)

